apps/aesophia/src/aeso_ast_infer_types.erl

```diff
@@ -1,256 +1,313 @@
+     {["Chain",    "gas_limit"],    Int}
+    ].
+
+infer([{contract, Attribs, ConName, Code}|Rest]) ->
     %% do type inference on each contract independently.
-    [{contract,Attribs,ConName,infer_contract(Code)}|infer(Rest)];
+    [{contract, Attribs, ConName, infer_contract(Code)}|infer(Rest)];
 infer([]) ->
     [].
 
 infer_contract(Defs) ->
     create_record_types(Defs),
-    C = unfold_record_types(infer_contract([],Defs)),
+    C = unfold_record_types(infer_contract(global_env(), Defs)),
     destroy_record_types(),
     C.
 
 %% infer_contract takes a proplist mapping global names to types, and
 %% a list of definitions.
-infer_contract(Env,[Def={type_def,_,_,_,_}|Rest]) ->
+infer_contract(Env, Defs) ->
+    Kind = fun({type_def, _, _, _, _})  -> type;
+              ({letfun, _, _, _, _, _}) -> function;
+              ({fun_decl, _, _, _})     -> prototype
+           end,
+    Get = fun(K) -> [ Def || Def <- Defs, Kind(Def) == K ] end,
     %% TODO: handle type defs
-    [Def|infer_contract(Env,Rest)];
-infer_contract(Env,[Fun={letfun,_,_,_,_,_}|Rest]) ->
-    {TypeSig,NewFun} = infer_nonrec(Env,Fun),
-    [NewFun|infer_contract([TypeSig|Env],Rest)];
-infer_contract(Env,[Rec={letrec,_,_}|Rest]) ->
-    {TypeSigs,NewRec} = infer_letrec(Env,Rec),
-    [NewRec|infer_contract(TypeSigs++Env,Rest)];
-infer_contract(_,[]) ->
-    [].
-
-infer_nonrec(Env,LetFun) ->
-    ets:new(type_vars,[set,named_table,public]),
-    create_unification_errors(),
-    create_field_constraints(),
-    NewLetFun = infer_letfun(Env,LetFun),
-    solve_field_constraints(),
-    Result = {TypeSig,_} = instantiate(NewLetFun),
-    destroy_and_report_unification_errors(),
-    destroy_field_constraints(),
-    ets:delete(type_vars),
-    print_typesig(TypeSig),
-    Result.
-
-infer_letrec(Env,{letrec,Attrs,Defs}) ->
-    ets:new(type_vars,[set,named_table,public]),
+    TypeDefs   = Get(type),
+    ProtoSigs = [ check_fundecl(Env, Decl) || Decl <- Get(prototype) ],
+    Env1      = ProtoSigs ++ Env,
+    {_, {letrec, _, Funs}} = infer_letrec(Env1, {letrec, [], Get(function)}),
+    TypeDefs ++ Funs.
+
+check_fundecl(_Env, {fun_decl, _Attrib, {id, _NameAttrib, Name}, {fun_t, _, Args, Ret}}) ->
+    {Name, {type_sig, Args, Ret}};  %% TODO: actually check that the type makes sense!
+check_fundecl(_, {fun_decl, _Attrib, {id, _, Name}, Type}) ->
+    error({fundecl_must_have_funtype, Name, Type}).
+
+%% infer_nonrec(Env, LetFun) ->
+%%     ets:new(type_vars, [set, named_table, public]),
+%%     create_unification_errors(),
+%%     create_field_constraints(),
+%%     NewLetFun = infer_letfun(Env, LetFun),
+%%     solve_field_constraints(),
+%%     Result = {TypeSig, _} = instantiate(NewLetFun),
+%%     destroy_and_report_unification_errors(),
+%%     destroy_field_constraints(),
+%%     ets:delete(type_vars),
+%%     print_typesig(TypeSig),
+%%     Result.
+
+infer_letrec(Env, {letrec, Attrs, Defs}) ->
+    ets:new(type_vars, [set, named_table, public]),
```

Comment:
Table is named: does it mean function is meant to be called by a process at a time? Does it mean this relies on conductor to build block candidate and verify imported blocks all in one thread?

Reply:
This is in the compiler. Run off chain. But a more functional approach would be nice.